feat: add push notifications

iamacook opened the pull request:
## What it solves

Implements push notifications for transaction events.

## How this PR fixes it

New "Notifications" settings sections have been added per-Safes/all added Safes, as well as a feature banner and associated tracking has been added. Subscribed categories dispatch push notifccations when the Safe UI is either out of focus or closed (but the browser is open).

The per-Safe "Notifications" section allows for notification subscription to all notification types, requiring a confirmatory signature as "Confirmation requests" are included. The global section is similar, subscribing to user selected (added) Safes with n number of signatures per selected chain.

Notifications are grouped into three categories: "Incoming transactions", "Outgoing transactions" and "Confirmation requests" according to [our Transaction Service webhooks](https://github.com/safe-global/safe-transaction-service/blob/master/safe_transaction_service/notifications/tasks.py).

- Incoming transactions
  - `INCOMING_ETHER`
  - `INCOMING_TOKEN`
- Outgoing transactions
  - `MODULE_TRANSACTION`
  - `EXECUTED_MULTISIG_TRANSACTION`
- Confirmation requests
  - `CONFIRMATION_REQUEST`

A banner is shown once per chain that has added Safes, subscribing to notifications for all added Safes on that chain.

User interaction within the UI is tracked directly. ~Due to technical limitations, direct notification interaction events (opening/closing) are cached and tracked when the UI is again opened.~ Moved to a separate PR https://github.com/safe-global/safe-wallet-web/pull/2500.

## How to test it

⚠️ Due to the complexity of this feature, I would advise explaining the feature on a call and maybe even testing it together. ⚠️

1. Registering for notifications should succeed, showing a notification and then displayind them accordingly when the Safe UI is not in focus or is closed (but the browser is open).
2. Unregistering a Safe should only unregister itself if there are others subscribed to on that chain or the entire device if it is the last Safe subscribed to on that chain.
3. Tracking of [user interaction](https://github.com/safe-global/safe-wallet-web/blob/ceaacff11584fbcebecfbd2f449c6a72b073825c/src/services/analytics/events/push-notifications.ts#L14-L95) should happen immediately.
4. ~Tacking of [notification iteraction](https://github.com/safe-global/safe-wallet-web/blob/ceaacff11584fbcebecfbd2f449c6a72b073825c/src/services/analytics/events/push-notifications.ts#L5-L13) should happen when the UI is reopened.~ (Moved to a separate PR https://github.com/safe-global/safe-wallet-web/pull/2500).
5. The MacOS info banner should only show on Mac devices.

If you want to reset your current registration, you need to:

- Reset browser notification permissions
- Unregister the `sw.js` service worker
- Delete all `firebase-*`/`notification-*` IndexedDB databases
![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/d65f4885-4b7b-474d-857e-d2680839caf2)

If you want to show the subscription banner again, you need to delete the `SAFE_v2__dismissPushNotifications` `localStorage` key.

## Screenshots

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/47e17311-e48b-43c9-82b5-87fd373a27c4)

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/58ef44f5-bccb-441c-a4fb-780a73dd8411)

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/34b0db2b-9882-47b0-8bec-097726d3a36e)

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/0d499dee-ec22-477c-862d-513b389f547f)

Please note that these are slightly dated but just demonstrate the notification. See the [designs](https://www.figma.com/file/ma5syWag1CO8yVbuzqkK0u/Push-notifications?mode=dev).

![notification](https://github.com/safe-global/safe-wallet-web/assets/20442784/2cb1f65a-58d3-41a5-89d1-817a19984fed)

## Checklist
* [ ] I've tested the branch on mobile 📱
* [x] I've documented how it affects the analytics (if at all) 📊
* [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

schmanu commented on src/components/settings/PushNotifications/logic.ts:
```diff
@@ -0,0 +1,118 @@
+}
+
+const getSafeRegistrationSignature = ({
+  safeAddresses,
```
Is there a case where we have more than one Safe address for the registration that requires a signature?

iamacook replied:
No. In the current design, you can only sign for notifications within a Safe.

schmanu commented on src/hooks/useFirebaseNotifications.ts:
```diff
@@ -0,0 +1,31 @@
+      return
+    }
+
+    initializeApp(FIREBASE_OPTIONS)
```
Can this throw if some constants are configured wrong / change? Should we try / catch the initialization?

iamacook replied:
Good point. I'll wrap it in a `try / catch`.

iamacook replied:
I abstracted the logic and used it in both places where we initialise Firebase in 6f106769e7298409835617aa88d7d5a1d794f346.

schmanu commented on src/services/firebase/index.ts:
```diff
@@ -0,0 +1,257 @@
+}
+
+const getLink = (path: string, query: { address: string; chain?: ChainInfo; safeTxHash?: string }) => {
+  const APP_URL = 'https://app.safe.global'
```
Can we somehow get the actual running app url? 
So localhost in dev / the PR deployments URLs?

schmanu commented on src/services/firebase/index.ts:
```diff
@@ -0,0 +1,257 @@
+    let balances: SafeBalanceResponse | undefined
+
+    try {
+      balances = await getBalances(chainId, address)
```
Just a thought:
On tx service side there is an endpoint to fetch token infos:
i.e. https://safe-transaction-goerli.safe.global/api/v1/tokens/0x61fD3b6d656F39395e32f46E2050953376c3f5Ff

Should we add this to the gateway service at some point? There are multiple examples where we fetch balances only to get some token info.

iamacook replied:
This sounds like a great idea. Let's file a feature request for it on the gateway.

iamacook commented on src/components/settings/PushNotifications/GlobalPushNotifications.tsx:
```diff
@@ -0,0 +1,327 @@
+        return acc
+      }, {})
+
+    const shouldRegisterSafes = Object.values(safesToRegister).some((safeAddresses) => safeAddresses.length > 0)
```
The former is relevant to (un-)selected Safes. I will rename them accordingly.

iamacook commented on src/components/settings/PushNotifications/GlobalPushNotifications.tsx:
```diff
@@ -0,0 +1,327 @@
+      return
+    }
+
+    // TODO: Enable when live on prod.
```
When I added this comment it wasn't released on the gateway. I'll investigate this now.

iamacook replied:
It has been since deployed. Added functionality in 6f106769e7298409835617aa88d7d5a1d794f346.

iamacook commented on public/firebase-messaging-sw.ts:
```diff
@@ -0,0 +1,74 @@
+
+  onBackgroundMessage(messaging, async (payload) => {
+    const ICON_PATH = '/images/safe-logo-green.png'
+    const DEFAULT_LINK = 'https://app.safe.global'
```
`self.location.origin` is now used from 6f106769e7298409835617aa88d7d5a1d794f346.

iamacook commented on src/components/settings/PushNotifications/GlobalPushNotifications.tsx:
```diff
@@ -0,0 +1,327 @@
+    })
+  }
+
+  // Wether Safes need to be (un-)reigstered with the service
```
Fixed in 6f106769e7298409835617aa88d7d5a1d794f346.

iamacook commented on src/components/settings/PushNotifications/GlobalPushNotifications.tsx:
```diff
@@ -0,0 +1,327 @@
+      }
+    }
+
+    Promise.all(registrationPromises)
```
Addressed in 6f106769e7298409835617aa88d7d5a1d794f346.

katspaugh commented on .github/workflows/build/action.yml:
```diff
@@ -43,3 +43,21 @@ runs:
         NEXT_PUBLIC_SAFE_RELAY_SERVICE_URL_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_SAFE_GELATO_RELAY_SERVICE_URL_STAGING }}
         NEXT_PUBLIC_IS_OFFICIAL_HOST: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_IS_OFFICIAL_HOST }}
         NEXT_PUBLIC_REDEFINE_API: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_REDEFINE_API }}
+        NEXT_PUBLIC_FIREBASE_API_KEY_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_API_KEY_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_DATABASE_URL_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_DATABASE_URL_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_PROJECT_ID_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_PROJECT_ID_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_APP_ID_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_APP_ID_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_VAPID_KEY_PRODUCTION: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_VAPID_KEY_PRODUCTION }}
+        NEXT_PUBLIC_FIREBASE_API_KEY_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_API_KEY_STAGING }}
+        NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN_STAGING }}
+        NEXT_PUBLIC_FIREBASE_DATABASE_URL_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_DATABASE_URL_STAGING }}
+        NEXT_PUBLIC_FIREBASE_PROJECT_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_PROJECT_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET_STAGING }}
+        NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_APP_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_APP_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_VAPID_KEY_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_VAPID_KEY_STAGING }}
```
That's a lot. Can we store them in a single env var as JSON?

katspaugh commented on next.config.mjs:
```diff
@@ -1,15 +1,6 @@
 import path from 'path'
+import { InjectManifest } from 'workbox-webpack-plugin'
 import withBundleAnalyzer from '@next/bundle-analyzer'
-import NextPwa from 'next-pwa'
-
-const withPWA = NextPwa({
-  disable: process.env.NODE_ENV === 'development',
-  dest: 'public',
-  reloadOnOnline: false,
-  /* Do not precache anything */
-  publicExcludes: ['**/*'],
-  buildExcludes: [/./],
-})
```
So there's no way to keep it?

katspaugh commented on src/components/settings/PushNotifications/GlobalPushNotifications.tsx:
```diff
@@ -0,0 +1,346 @@
+
+// Convert data structure of added Safes
+export const transformAddedSafes = (addedSafes: AddedSafesState): NotifiableSafes => {
+  const obj: NotifiableSafes = {}
+
+  for (const [chainId, addedSafesOnChain] of Object.entries(addedSafes)) {
+    obj[chainId] = Object.keys(addedSafesOnChain)
+  }
+
+  return obj
+}
+
+// Convert data structure of currently notified Safes
+const transformCurrentSubscribedSafes = (allPreferences: NotificationPreferences): NotifiableSafes => {
+  const obj: NotifiableSafes = {}
+
+  for (const { chainId, safeAddress } of Object.values(allPreferences)) {
+    if (!obj[chainId]) {
+      obj[chainId] = []
+    }
+
+    obj[chainId].push(safeAddress)
+  }
+
+  return obj
```
In both of these a `reduce` would be a little easier to follow.

iamacook replied:
Adjusted accordingly in 75b5a63d8189dbe37b42ffb469b1bb2361a2d28a.

katspaugh commented on src/components/settings/PushNotifications/GlobalPushNotifications.tsx:
```diff
@@ -0,0 +1,346 @@
+    if (shouldRegister) {
+      registrationPromises.push(registerNotifications(safesToRegister))
+    }
+
+    const shouldUnregister =
+      currentNotifiedSafes &&
+      Object.entries(currentNotifiedSafes).reduce<NotifiableSafes>((acc, [chainId, safeAddresses]) => {
+        const safesToUnregisterOnChain = safeAddresses.filter(
+          (safeAddress) => !selectedSafes[chainId]?.includes(safeAddress),
+        )
+
+        if (safesToUnregisterOnChain.length > 0) {
+          acc[chainId] = safeAddresses
+        }
+        return acc
+      }, {})
+
+    if (shouldUnregister) {
+      for (const [chainId, safeAddresses] of Object.entries(shouldUnregister)) {
+        const shouldUnregsiterDevice =
+          safeAddresses.length === currentNotifiedSafes[chainId].length &&
+          safeAddresses.every((safeAddress) => {
+            return currentNotifiedSafes[chainId]?.includes(safeAddress)
+          })
+
+        if (shouldUnregsiterDevice) {
+          registrationPromises.push(unregisterChainNotifications(chainId))
+          continue
+        }
+
+        for (const safeAddress of safeAddresses) {
+          registrationPromises.push(unregisterSafeNotifications(chainId, safeAddress))
+        }
+      }
+    }
+
+    await Promise.all(registrationPromises)
+
+    trackEvent(PUSH_NOTIFICATION_EVENTS.SAVE_SETTINGS)
+  }
```
A lot is going on here and it's a bit hard to follow. Especially with regards to what functions need the state of the component and which are pure. I would suggest to extract as much logic as you can into pure functions outside of the component and keep JSX separate from logic.

iamacook replied:
I've abstracted the code outside of the component in 75b5a63d8189dbe37b42ffb469b1bb2361a2d28a.

katspaugh reviewed (commented):
It looks great now in terms of what goes into the Service Worker. We could add a comment on top of each module used in the SW, saying something to the effect of "be careful what you import here".

iamacook commented:
>We could add a comment on top of each module used in the SW, saying something to the effect of "be careful what you import here".

I've added comments in the relevant places.

> Also, you're probably aware, but there's now no tracking of clicks on notifications now (I think you had them in the previous version).

Now that I've addresses your review, I will now look our proposed solution of caching the events and dispatching them when the UI is opened.

iamacook commented on src/tests/pages/apps-share.test.tsx:
```diff
@@ -4,16 +4,67 @@ import ShareSafeApp from '@/pages/share/safe-app'
 import * as useWalletHook from '@/hooks/wallets/useWallet'
 import * as useOwnedSafesHook from '@/hooks/useOwnedSafes'
+import * as manifest from '@/services/safe-apps/manifest'
+import * as sdk from '@safe-global/safe-gateway-typescript-sdk'
```
This test was breaking and I realised that none of the responses were mocked.

katspaugh commented on src/services/firebase/notifications.ts:
```diff
@@ -0,0 +1,245 @@
+  if (isWebhookEvent(payload.data)) {
+    const webhookNotification = await _parseWebhookNotification(payload.data)
+
+    if (webhookNotification) {
+      return {
+        title: webhookNotification.title,
+        body: webhookNotification.body,
+        link: webhookNotification.link,
+      }
+    }
+  }
+
+  // Manually dispatched notifications from the Firebase admin panel
+  // Displayed as is
+  if (payload.notification) {
+    return {
+      title: payload.notification.title || '',
+      body: payload.notification.body,
+      image: payload.notification.image,
+    }
+  }
```
I feel like this could be condensed into two-three lines?

iamacook replied:
Addressed in bbe73f4277ad22d090b588b66f45f57e3d06d560.

katspaugh commented on src/components/settings/PushNotifications/logic.ts:
```diff
@@ -0,0 +1,121 @@
+  try {
+    permission = await Notification.requestPermission()
+  } catch (e) {
+    console.error('Error requesting notification permission', e)
```
Any reason not to use the logError function here? Is this module running in the SW?

iamacook replied:
Addressed in bbe73f4277ad22d090b588b66f45f57e3d06d560.

katspaugh reviewed (commented):
I'm lacking clarity on which modules are used by a service worker and which are not. Can we either name them in a distinct way or maybe add a comment on top of each?

iamacook commented on src/services/firebase/notifications.ts:
```diff
@@ -0,0 +1,245 @@
+
+  const tokenInfo = await getBalances(chainId, safeAddress, DEFAULT_CURRENCY)
+    .then(({ items }) => items.find((token) => token.tokenInfo.address === tokenAddress)?.tokenInfo)
+    .catch(() => null)
```
Addressed in bbe73f4277ad22d090b588b66f45f57e3d06d560.

iamacook commented:
> I'm lacking clarity on which modules are used by a service worker and which are not. Can we either name them in a distinct way or maybe add a comment on top of each?

I've restructured the code, moving that which is relevant _only_ for the service worker to the `@/service-workers/firebase` folder, as well as rename functions accordingly in bbe73f4277ad22d090b588b66f45f57e3d06d560.

katspaugh commented on src/service-workers/firebase-messaging/notifications.ts:
```diff
@@ -0,0 +1,240 @@
+  return preferencesStore.preferences[type]
+}
+
+const GATEWAY_URL_PRODUCTION = process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'https://safe-client.safe.global'
+const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'https://safe-client.staging.5afe.dev'
+
+// localStorage cannot be accessed in service workers so we reference the flag from the environment
+const GATEWAY_URL = FIREBASE_IS_PRODUCTION ? GATEWAY_URL_PRODUCTION : GATEWAY_URL_STAGING
+setBaseUrl(GATEWAY_URL)
```
Could you move this to the top of the file?

iamacook replied:
Addressed in a115c14973ce9cf0f7b6af14d7ae633ef6af035d.

katspaugh commented on src/service-workers/firebase-messaging/notifications.ts:
```diff
@@ -0,0 +1,240 @@
+    },
+    [WebhookType.INCOMING_TOKEN]: async ({ address, txHash, tokenAddress, value }) => {
+      const token = await getTokenInfo(data.chainId, address, tokenAddress, value)
+      return {
+        title: `${token.name} received`,
+        body: `Safe ${shortenAddress(address)} on ${chainName} received ${token.value} ${
+          token.symbol
+        } in transaction ${shortenAddress(txHash)}.`,
+      }
+    },
+    [WebhookType.OUTGOING_TOKEN]: async ({ address, txHash, tokenAddress, value }) => {
+      const token = await getTokenInfo(data.chainId, address, tokenAddress, value)
+      return {
+        title: `${token.name} sent`,
+        body: `Safe ${shortenAddress(address)} on ${chainName} sent ${token.value} ${
+          token.symbol
+        } in transaction ${shortenAddress(txHash)}.`,
+      }
+    },
+    [WebhookType.MODULE_TRANSACTION]: ({ address, module, txHash }) => {
+      return {
+        title: 'Module transaction',
+        body: `Safe ${shortenAddress(address)} on ${chainName} executed a module transaction ${shortenAddress(
+          txHash,
+        )} from module ${shortenAddress(module)}.`,
+      }
+    },
+    [WebhookType.CONFIRMATION_REQUEST]: ({ address, safeTxHash }) => {
+      return {
+        title: 'Confirmation request',
+        body: `Safe ${shortenAddress(
+          address,
+        )} on ${chainName} has a new confirmation request for transaction ${shortenAddress(safeTxHash)}.`,
+      }
+    },
+    [WebhookType.SAFE_CREATED]: () => {
+      // We do not preemptively subscribe to Safes before they are created
+      return null
+    },
+  }
```
This could be moved to a separate module as a static config.

iamacook replied:
Moved in a115c14973ce9cf0f7b6af14d7ae633ef6af035d.

usame-algan commented on src/components/settings/PushNotifications/logic.ts:
```diff
@@ -0,0 +1,123 @@
+
+  trackEvent(isGranted ? PUSH_NOTIFICATION_EVENTS.GRANT_PERMISSION : PUSH_NOTIFICATION_EVENTS.REJECT_PERMISSION)
+
+  if (!isGranted) {
+    alert('You must allow notifications to register your device.')
+  }
```
We show this alert on the global preferences but not when on a specific safe. Is there a reason for that?

iamacook replied:
I've removed this. The function is now being used as an early return and I've added a comment explaining it in 12116b71f0067a697e64c861fc682219a9014108.

usame-algan commented on src/components/settings/PushNotifications/logic.ts:
```diff
@@ -0,0 +1,123 @@
+  const app = initializeFirebaseApp()
+  const messaging = getMessaging(app)
+
+  const token = await getToken(messaging, {
```
The documentation of this function says: `If a notification permission isn't already granted, this method asks the user for permission.`

Isn't that what we also do manually in `requestNotificationPermission`?

iamacook replied:
Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

iamacook replied:
See above comment.

usame-algan commented on src/hooks/useDecodeTx.ts:
```diff
@@ -14,7 +14,7 @@ const useDecodeTx = (tx?: SafeTransaction): AsyncResult<DecodedDataResponse> =>
 
   const [data = nativeTransfer, error, loading] = useAsync<DecodedDataResponse>(() => {
-    if (!encodedData || isEmptyData) return
+    if (!encodedData || isEmptyData || !tx.data.to) return
```
The parameter should be optional now so I think we can remove this check.

iamacook replied:
Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

iamacook commented on package.json:
```diff
@@ -118,14 +120,17 @@
     "typescript": "4.9.4",
-    "typescript-plugin-css-modules": "^4.2.2"
+    "typescript-plugin-css-modules": "^4.2.2",
+    "webpack": "^5.88.2"
```
It is, but [`next-pwa` requires it](https://ducanh-next-pwa.vercel.app/docs/next-pwa/getting-started#install) to build service workers.

iamacook commented on src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx:
```diff
@@ -0,0 +1,137 @@
+      return
+    }
+
+    trackEvent(PUSH_NOTIFICATION_EVENTS.DISPLAY_BANNER)
```
Iirc, when I went over the required events with @kirkkonen, this was added to ease funnel creation.

It cannot be moved to `dismissBanner` we show the banner immediately.

usame-algan replied:
Maybe to track how many users see the banner but don't dismiss it? But since the banner is closed when clicking anywhere I don't think that would be a useful metric.

iamacook replied:
As per the addition of a close button, this is now not relevant.

iamacook commented on src/components/settings/PushNotifications/index.tsx:
```diff
@@ -0,0 +1,263 @@
+                  control={
+                    <Checkbox
+                      checked={preferences[WebhookType.NEW_CONFIRMATION]}
+                      onChange={(_, checked) => {
+                        setPreferences({
+                          ...preferences,
+                          [WebhookType.NEW_CONFIRMATION]: checked,
+                        })
```
Yes. As `indexedDB` is promised base, we're only changing the state after it has resolved. We could change this to be optimistic but I think having the database as the SSOT makes most sense. Wdyt?

usame-algan replied:
Agreed, lets not make it optimistic but maybe we can provide some loading indicator to the user.

iamacook commented on src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts:
```diff
@@ -0,0 +1,217 @@
+
+    clearIndexedDb(preferencesStore)
+      .then(hydratePreferences)
+      .catch(() => null)
```
We only need to catch errors, the returned value isn't needed. However, as we have the `no-empty` TSLint rule on, an "empty" function won't work. Would you prefer I change them to `try / catch`?

usame-algan replied:
I think the current style is elegant but I am not sure if we should log anything in case those indexed db calls throw. I will do some research on the topic.

iamacook commented on src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx:
```diff
@@ -0,0 +1,137 @@
+    setDismissedBannerPerChain({
+      ...dismissedBannerPerChain,
+      [safe.chainId]: true,
+    })
```
Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

iamacook commented on src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx:
```diff
@@ -0,0 +1,137 @@
+    })
+  }, [dismissedBannerPerChain, safe.chainId, setDismissedBannerPerChain])
+
+  // Click outside to dismiss banner
```
Good point. We're using the same style as the batch banner atm. @liliiaorlenko, what are your thoughts?

iamacook replied:
As per discussion, I've added a close button to the banner which also negates the above comment regarding the tracking.

![image](https://github.com/safe-global/safe-wallet-web/assets/20442784/b78d006a-e337-47fc-8696-ff4098476120)

iamacook commented on src/components/settings/PushNotifications/index.tsx:
```diff
@@ -0,0 +1,263 @@
+  }
+
+  const isMac = typeof navigator !== 'undefined' && navigator.userAgent.includes('Mac')
+  const shouldShowMacHelper = isMac || IS_DEV
```
I've created a new `useIsMac` hook in line with NextJS' usage of the `window` object in 12116b71f0067a697e64c861fc682219a9014108.

iamacook commented on src/components/settings/PushNotifications/index.tsx:
```diff
@@ -0,0 +1,263 @@
+                                variant: 'success',
+                                groupKey: 'notifications',
+                              }),
+                            )
```
Fixed in 12116b71f0067a697e64c861fc682219a9014108.

iamacook commented on src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  // Delete all preferences store entries
+  const clearPreferences = () => {
```
Renamed in 12116b71f0067a697e64c861fc682219a9014108.

iamacook commented on src/components/settings/PushNotifications/PushNotificationsBanner/styles.module.css:
```diff
@@ -0,0 +1,54 @@
+  font-size: 12px;
+  width: var(--space-5);
+  height: 24px;
+  z-index: 9999999;
```
Indeed. Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

usame-algan reviewed (approved):
Works great and thanks for implementing the changes 👍

Will discuss the notification dismissal tomorrow but it is non-blocking imo. LGTM 🚀

liliya-soroka commented:
Feedback: A notification banner is displayed for the safe with enabled notifications through Global settings. It's confusing, as I just enabled and signed notifications enabling for this safe 
<img width="525" alt="image" src="https://github.com/safe-global/safe-wallet-web/assets/338622/c24a8084-ab41-449b-a06e-5bf068050534">
Steps:
1. make sure that you have a few safes in the app on different networks
2. run app 
3. Select Enable Notifications using Global settings
4. enable notifications for all your safes 
5. come back to the safes list and switch to the safe on another network
Current result: Notifications banner is displayed

francovenica commented:
Looks good.
The ticket  was checked by several people, in windows and mac for the different type of notifications. 
All the issues reported in the QA document were fixed or discussed: https://www.notion.so/safe-global/QA-Feedback-for-notifications-epic-2391-557125a8a91240a59903655e0cfc3bd3

A few notes for future iterations that we might want to revisit:
You get a notification for every tx executed in a "bulk transaction execution".
The "Confirm request" only works for safes with at least 3 owners and a policy of at least 2. I assume this is cuz for a threshold of 1 you don't need to confirm anymore, and if the safe has 2 owners, the tx 2nd confirmation needed is a "sign and execute" which I believe is different from a "just confirm" type of signing.
